api/src/main/java/org/openmrs/module/appframework/service/AppFrameworkServiceImpl.java

```diff
@@ -322,60 +314,84 @@ && hasPrivilege(userContext, candidate.getRequiredPrivilege())) {
+			script.append(requireExpressionScripts.get(key));
+			script.append(System.lineSeparator());
+		}
+		script.append(System.lineSeparator());
+
+		// If any properties in contextModel are Maps, we want to allow scripts to access their subproperties
+		// with dot notation, but ScriptEngine and Bindings don't naturally handle this. Instead we will convert
+		// all Map-type properties to JSON and use this to define objects directly within the ScriptEngine.
+		// (Properties that are not Maps don't need this special treatment.)
+		ObjectMapper objectMapper = new ObjectMapper();
+		Map<String, Object> mapProperties = new HashMap<>();
+		for (Map.Entry<String, Object> e : contextModel.entrySet()) {
+			if (e.getValue() instanceof Map) {
+				mapProperties.put(e.getKey(), e.getValue());
+			}
+			else if(!"util".equals(e.getKey())) {
+				script.append("var ").append(e.getKey()).append(" = ");
+				script.append(objectMapper.writeValueAsString(e.getValue())).append(";");
+				script.append(System.lineSeparator());
+			}
+		}
+
+		for (Map.Entry<String, Object> e : mapProperties.entrySet()) {
+			try {
+				script.append("var ").append(e.getKey()).append(" = ");
+				script.append(objectMapper.writeValueAsString(e.getValue())).append(";");
+				script.append(System.lineSeparator());
+			}
+			catch (Exception ex) {
+				StringBuilder extraInfo = new StringBuilder();
+				extraInfo.append("type:").append(e.getValue().getClass().getName());
+				if (e.getValue() instanceof Map) {
+					Map<?, ?> map = (Map<?, ?>) e.getValue();
+					extraInfo.append(" properties:");
+					for (Map.Entry<?, ?> entry : map.entrySet()) {
+						extraInfo.append(" ").append(entry.getKey().toString()).append(":");
+						extraInfo.append(entry.getValue() == null ? "null" : entry.getValue().getClass().toString());
 					}
 				}
-
-				return javascriptEngine.eval("(" + requireExpression + ") == true").equals(Boolean.TRUE);
+				throw new IllegalStateException("Failed to set '" + e.getKey() + "' variable (" + extraInfo + ")", ex);
 			}
+		}
+		script.append(System.lineSeparator());
+		return script.toString();
+	}
 
+	@Override
+	public boolean checkRequireExpression(Requireable candidate, AppContextModel contextModel) {
+		try {
+			return checkRequireExpressionStrict(candidate, contextModel);
 		}
 		catch (Exception e) {
-			log.error("Failed to evaluate 'require' check for extension " + candidate.getId(), e);
+			log.error("Failed to evaluate 'require' check for extension " + candidate.getId());
 			return false;
 		}
 	}
 
+	@Override
+	public boolean checkRequireExpressionStrict(Requireable candidate, AppContextModel contextModel) throws Exception {
+		String requireExpression = candidate.getRequire();
+		if (StringUtils.isBlank(requireExpression)) {
+			return true;
+		}
+		else {
+			String script = getRequireExpressionContext(contextModel) + System.lineSeparator() + requireExpression;
+			return javascriptEngine.eval(script).equals(Boolean.TRUE);
+		}
+	}
```

Comment:
This simply adds a new method that mirrors the existing one, but which does not swallow exceptions.  The existing method delegates to this and handles the exception to retain backwards compatibility.